### PR TITLE
SCP-3673: SCP-2997: versioned builtins. fix consByteString integer wrapping

### PR DESCRIPTION
**Pull request by bezirg**
This PR contains two patches:

1) to attach a PLC.Version to builtins, so that it can be interpreted in `toBuiltinMeaning`
2) fix `consByteString` by changing its type to `Word8 -> ByteString -> ByteString` and let the `readKnown` handle the type-conversion and failure on number wrapping.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs [hunk 1]
```diff
@@ -334,3 +335,13 @@ instance uni ~ DefaultUni => ToBuiltinMeaning uni ExtensionFun where
...
             let uniList' = DefaultUniList $ DefaultUniPair DefaultUniBool uniA
             pure . SomeConstant . someValueOf uniList' $ map swap xs
+
+    toBuiltinMeaning (Version _ majorNum _ _) MajorVersion =
+        makeBuiltinMeaning
+        @(() -> EvaluationResult Integer)
+        (\(_ :: ()) -> case majorNum of
+                1 -> EvaluationSuccess 1
+                2 -> EvaluationSuccess 2
+                _ -> EvaluationFailure
```
Comment (effectfully) [inline plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:345]: Great test!

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs [hunk 2]
```diff
@@ -119,6 +119,7 @@ data ExtensionFun
     | Swap  -- For checking that permuting type arguments of a polymorphic built-in works correctly.
     | SwapEls  -- For checking that nesting polymorphic built-in types and instantiating them with
                -- a mix of monomorphic types and type variables works correctly.
+    | MajorVersion -- reflect the major "plutu version number" of the program
```
Comment (effectfully) [inline plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:122]:
```suggestion
    | MajorVersion -- reflect the major "plutu version number" of the program
```

**Review by effectfully — COMMENTED** [review]
The slowdown is +1.11% on average. I do not see how it can be any slower, given that the meanings of builtins are cached, so the overhead has to be negligible and from what I remember we construct the meanings of builtins outside of benchmarking.

File: plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs (hunk 2 above)
Comment (effectfully) [reply plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:122]: Oops, sorry. Let me try again.

Comment (effectfully) [reply plutus-core/plutus-core/examples/PlutusCore/Examples/Builtins.hs:122]:
```suggestion
    | MajorVersion  -- Reflect the major "Plutus version number" of the program.
```

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs [hunk 3]
```diff
@@ -71,11 +71,11 @@ class (Typeable uni, Typeable fun, Bounded fun, Enum fun, Ix fun) => ToBuiltinMe
     type CostingPart uni fun
 
     -- | Get the 'BuiltinMeaning' of a built-in function.
-    toBuiltinMeaning :: HasConstantIn uni val => fun -> BuiltinMeaning val (CostingPart uni fun)
+    toBuiltinMeaning :: HasConstantIn uni val => Version ann -> fun -> BuiltinMeaning val (CostingPart uni fun)
```
Comment (effectfully) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: We could move `Version` to the type level and add it as an argument to `ToBuiltinMeaning`, so that we don't need to bind it in each builtin. We could also index `DefaultFun` by `ver` and avoid extending `ToBuiltinMeaning` in any way at all. In both the cases we'd need to demote `ver` to a `Version` via a singleton or whatever, but that should be easy.

None of the options are probably worth it, though.

**Comment by effectfully** [comment]
> he slowdown is +1.11% on average. I do not see how it can be any slower

I suggest taking a look at the generated Core, maybe there's something obvious in there. If that's not the case, I'll take a look myself, since I know how Core is supposed to look for builtins.

For checking add

```
{-# OPTIONS_GHC -ddump-simpl -ddump-to-file -dsuppress-all -fforce-recomp -dumpdir /tmp/dumps #-}
```

to `EvaluationContext.hs` and check how one of the `ToBuiltinMeaning` instances looks like. In particular how the second argument of a `BuiltinMeaning` gets compiled.

File: plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs (hunk 3 above)
Comment (michaelpj) [inline plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: I think we should not use the `Version` type for this. It's not 100% clear what it means, but I think it should refer to the version of plutus core the _language_, which is naturally independent of the set of builtins. Yes, this is hopelessly confused by the fact that that ledger also talks about plutus language versions, but tough.

Maybe a new type in `Default/Builtins.hs`, `BuiltinsVersion` or something? Because that's actually the thing we're changing. And then in the V2 API we would say we're using the V2 builtins.

The annoying thing is that we need a type here in this typeclass method. I guess we could stick an associated type in there, so we have something like:
```
class ... ToBuiltinMeaning uni fun where
   type BuiltinVersion uni fun
   toBuiltinMeaning :: HasConstantIN uni val => BuiltinVersion fun -> BuiltinMeaning val (CostingPart uni fun)
```

Comment (michaelpj) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: Yeah, although I thought you were mostly talking about pushing things into the types. I mostly just wanted us to not reuse `Version`, I don't have so much opinion about whether we should do it at the value or the type level.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: Or we could simply have 

```haskell
data Versioned v a = Versioned !v !a

data DefaultFunVersion
    = DefaultFunV1
    | DefaultFunV2

instance uni ~ DefaultUni => ToBuiltinMeaning uni (Versioned DefaultFunVersion DefaultFun) where
    type CostingPart uni (Versioned DefaultFunVersion DefaultFun) = BuiltinCostModel
```

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: Also doesn't force us to explicitly ignore a redundant argument in every builtin other than `ConsByteString`. Doesn't impose any versioning on sets of builtins where we don't care about versioning, like the `NopFun` and `ExtensionFun` sets. And may also fix the performance problem right away.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: > Right, but then we'd have to actually put those `Versioned` constructors in the program everywhere, right?

No, you can always recover `toBuiltinMeaning :: DefaultFunVersion -> DefaultFun -> blah-blah-blah` from the `Versioned`-based interface by simply packaging the version with the builtin using the `Versioned` constructor.

Comment (effectfully) [reply plutus-core/plutus-core/src/PlutusCore/Builtin/Meaning.hs:74]: I.e.  `f :: DefaultFunVersion -> DefaultFun -> r` is the same as `f :: Versioned DefaultFunVersion DefaultFun -> r`, except the latter is more convenient as it doesn't require changing as much machinery and imposing versioning on everything.
